---
title: "03 16s Analysis Report"
output:
  pdf_document:
    df_print: kable
  html_document:
    df_print: kable
    smart: false
---

```{r init renv, message=FALSE, warning=FALSE, include=FALSE}
library(renv)
#renv::isolate()
#renv::settings$snapshot.type("all")
renv::restore()
```

```{r parse item of interest and make outdir for figures, echo=FALSE, message=FALSE, warning=FALSE}
#this may be required for each file
#item of interest, comes from bash and python script that orchestrates everything
wd <- getwd()
ioi <- read.csv("../item_of_interest.csv")
ioi<-as.character(ioi)

ioi_ord <- read.csv("../order_item_of_interest.csv")
colnames(ioi_ord) <- ioi
ioi_ord <- as.list(ioi_ord)[[ioi]]

outdir_name <- "Figures"
if(!dir.exists(paste0("../",outdir_name))) {dir.create(paste0("../",outdir_name))}
outdir <- paste0("../",outdir_name)
```

```{r ampvis code from github so we dont need to query, echo=FALSE, message=FALSE, warning=FALSE}
phyloseq_to_ampvis2 <- function(physeq) {
  #check object for class
  if(!any(class(physeq) %in% "phyloseq"))
    stop("physeq object must be of class \"phyloseq\"", call. = FALSE)
  
  #ampvis2 requires taxonomy and abundance table, phyloseq checks for the latter
  if(is.null(physeq@tax_table))
    stop("No taxonomy found in the phyloseq object and is required for ampvis2", call. = FALSE)
  
  #OTUs must be in rows, not columns
  if(phyloseq::taxa_are_rows(physeq))
    abund <- as.data.frame(phyloseq::otu_table(physeq)@.Data)
  else
    abund <- as.data.frame(t(phyloseq::otu_table(physeq)@.Data))
  
  #tax_table is assumed to have OTUs in rows too
  tax <- phyloseq::tax_table(physeq)@.Data
  
  #merge by rownames (OTUs)
  otutable <- merge(
    abund,
    tax,
    by = 0,
    all.x = TRUE,
    all.y = FALSE,
    sort = FALSE
  )
  colnames(otutable)[1] <- "OTU"
  
  #extract sample_data (metadata)
  if(!is.null(physeq@sam_data)) {
    metadata <- data.frame(
      phyloseq::sample_data(physeq),
      row.names = phyloseq::sample_names(physeq), 
      stringsAsFactors = FALSE, 
      check.names = FALSE
    )
    
    #check if any columns match exactly with rownames
    #if none matched assume row names are sample identifiers
    samplesCol <- unlist(lapply(metadata, function(x) {
      identical(x, rownames(metadata))}))
    
    if(any(samplesCol)) {
      #error if a column matched and it's not the first
      if(!samplesCol[[1]])
        stop("Sample ID's must be in the first column in the sample metadata, please reorder", call. = FALSE)
    } else {
      #assume rownames are sample identifiers, merge at the end with name "SampleID"
      if(any(colnames(metadata) %in% "SampleID"))
        stop("A column in the sample metadata is already named \"SampleID\" but does not seem to contain sample ID's", call. = FALSE)
      metadata$SampleID <- rownames(metadata)
      
      #reorder columns so SampleID is the first
      metadata <- metadata[, c(which(colnames(metadata) %in% "SampleID"), 1:(ncol(metadata)-1L)), drop = FALSE]
    }
  } else
    metadata <- NULL
  
  #extract phylogenetic tree, assumed to be of class "phylo"
  if(!is.null(physeq@phy_tree)) {
    tree <- phyloseq::phy_tree(physeq)
  } else
    tree <- NULL
  
  #extract OTU DNA sequences, assumed to be of class "XStringSet"
  if(!is.null(physeq@refseq)) {
    #convert XStringSet to DNAbin using a temporary file (easiest)
    fastaTempFile <- tempfile(pattern = "ampvis2_", fileext = ".fa")
    Biostrings::writeXStringSet(physeq@refseq, filepath = fastaTempFile)
  } else
    fastaTempFile <- NULL
  
  #load as normally with amp_load
  ampvis2::amp_load(
    otutable = otutable,
    metadata = metadata,
    tree = tree,
    fasta = fastaTempFile
  )
}
```

## 3 Species Abundance Heatmap

```{r heatmap at the Genus level annotatted with item of interest, echo=FALSE, fig.height=14, fig.width=28, message=FALSE, warning=FALSE}
#original height 10 width 14
library(qiime2R)
library(phyloseq)
library(ampvis2)
library(ggplot2)

table_dada2 <- "../qiime/core-metric-results/rarefied_table.qza"
rooted_tree <- "../qiime/rooted-tree.qza"
taxonomy_file <- "../qiime/taxonomy.qza"
metadata_file <- "../metadata.tsv"

#devtools::source_gist("8d0ca4206a66be7ff6d76fc4ab8e66c6")
cycle_1 <- qza_to_phyloseq(table_dada2,rooted_tree,taxonomy_file,metadata_file) 
cycle_1@sam_data[[ioi]] <- factor(cycle_1@sam_data[[ioi]], levels= ioi_ord)

cycle_1_amp <- phyloseq_to_ampvis2(cycle_1)

heatmap <- ampvis2::amp_heatmap(cycle_1_amp,
                     group_by = "SampleID", 
                     tax_aggregate = "Genus",
                     tax_show = 30,
                     facet_by = ioi,
                     tax_add = "Phylum",
                     plot_values = FALSE,
                     plot_colorscale = "log10") +
  theme(text=element_text(size = 20))

heatmap
ggsave("heatmap.png", plot=heatmap, path=outdir)

```


```{r read in cutoffs used by qiime, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(kableExtra)
cutoffs = read.csv("../qiime/cutoffs.csv",sep = ',',strip.white = TRUE)
#cutoffs = knitr::kable(cutoffs)
cutoffs %>% kbl() %>% kable_styling(bootstrap_options = c("striped","condensed","responsive"))
```


```{r make data table for dada2 stats from qiime, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(kableExtra)
dadastat = read.csv("../dada2_stats.tsv",sep = '\t',strip.white = TRUE)
#dadastat_t = dadastat %>% kbl() %>% kable_classic(full_width = F)
dadastat %>% kbl() %>% kable_styling(bootstrap_options = c("striped","condensed","responsive")) 
```

```{r table for sampling depth in qiime, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(kableExtra)
depth = read.csv("../qiime/sampling_depth.csv", sep=",")
depth %>% kbl() %>% kable_styling(bootstrap_options = c("striped","condensed","responsive"))
```