---
title: "08 16s Analysis Report"
output:
  pdf_document:
    df_print: kable
  html_document:
    df_print: kable
    smart: false
---

```{r init renv, message=FALSE, warning=FALSE, include=FALSE}
#library(renv)
#renv::isolate()
#renv::settings$snapshot.type("all")
#renv::restore()
```

```{r parse item of interest and make outdir for figures, echo=FALSE, message=FALSE, warning=FALSE}
#this may be required for each file
#item of interest, comes from bash and python script that orchestrates everything
wd <- getwd()
ioi <- readLines("item_of_interest.csv")
ioi<-as.character(ioi)

ioi_ord <- read.csv("order_item_of_interest.csv")
colnames(ioi_ord) <- ioi
ioi_ord <- as.list(ioi_ord)[[ioi]]

outdir_name <- "Figures"
if(!dir.exists(outdir_name)) {dir.create(outdir_name)}
outdir <- outdir_name

# library(tidyverse)
# library(reshape2)
# library(dplyr)
# library(ggplot2)
```
```{r ampvis code from github so we dont need to query, echo=FALSE, message=FALSE, warning=FALSE}
phyloseq_to_ampvis2 <- function(physeq) {
  #check object for class
  if(!any(class(physeq) %in% "phyloseq"))
    stop("physeq object must be of class \"phyloseq\"", call. = FALSE)
  
  #ampvis2 requires taxonomy and abundance table, phyloseq checks for the latter
  if(is.null(physeq@tax_table))
    stop("No taxonomy found in the phyloseq object and is required for ampvis2", call. = FALSE)
  
  #OTUs must be in rows, not columns
  if(phyloseq::taxa_are_rows(physeq))
    abund <- as.data.frame(phyloseq::otu_table(physeq)@.Data)
  else
    abund <- as.data.frame(t(phyloseq::otu_table(physeq)@.Data))
  
  #tax_table is assumed to have OTUs in rows too
  tax <- phyloseq::tax_table(physeq)@.Data
  
  #merge by rownames (OTUs)
  otutable <- merge(
    abund,
    tax,
    by = 0,
    all.x = TRUE,
    all.y = FALSE,
    sort = FALSE
  )
  colnames(otutable)[1] <- "OTU"
  
  #extract sample_data (metadata)
  if(!is.null(physeq@sam_data)) {
    metadata <- data.frame(
      phyloseq::sample_data(physeq),
      row.names = phyloseq::sample_names(physeq), 
      stringsAsFactors = FALSE, 
      check.names = FALSE
    )
    
    #check if any columns match exactly with rownames
    #if none matched assume row names are sample identifiers
    samplesCol <- unlist(lapply(metadata, function(x) {
      identical(x, rownames(metadata))}))
    
    if(any(samplesCol)) {
      #error if a column matched and it's not the first
      if(!samplesCol[[1]])
        stop("Sample ID's must be in the first column in the sample metadata, please reorder", call. = FALSE)
    } else {
      #assume rownames are sample identifiers, merge at the end with name "SampleID"
      if(any(colnames(metadata) %in% "SampleID"))
        stop("A column in the sample metadata is already named \"SampleID\" but does not seem to contain sample ID's", call. = FALSE)
      metadata$SampleID <- rownames(metadata)
      
      #reorder columns so SampleID is the first
      metadata <- metadata[, c(which(colnames(metadata) %in% "SampleID"), 1:(ncol(metadata)-1L)), drop = FALSE]
    }
  } else
    metadata <- NULL
  
  #extract phylogenetic tree, assumed to be of class "phylo"
  if(!is.null(physeq@phy_tree)) {
    tree <- phyloseq::phy_tree(physeq)
  } else
    tree <- NULL
  
  #extract OTU DNA sequences, assumed to be of class "XStringSet"
  if(!is.null(physeq@refseq)) {
    #convert XStringSet to DNAbin using a temporary file (easiest)
    fastaTempFile <- tempfile(pattern = "ampvis2_", fileext = ".fa")
    Biostrings::writeXStringSet(physeq@refseq, filepath = fastaTempFile)
  } else
    fastaTempFile <- NULL
  
  #load as normally with amp_load
  ampvis2::amp_load(
    otutable = otutable,
    metadata = metadata,
    tree = tree,
    fasta = fastaTempFile
  )
}
```

## 8 Ranked Abundance Curves

```{r, echo=FALSE, fig.height=10, fig.width=16, message=FALSE, warning=FALSE}
#library(devtools)
#llibrary(remotes)
#install.packages("BiocManager")
#BiocManager::install(version="3.13")
library(ampvis2)
#b655a307c8431201573096ab57140b46d8733dd7
library(ggplot2)
library(qiime2R)
#d1ad96657ada993cf6c2841b29113a4f635c6b56
library(phyloseq)
library(tidyverse)
#@1.3.1

#table_dada2 <- "table-dada2.qza"
table_dada2 <- "core-metric-results/rarefied_table.qza"
rooted_tree <- "rooted-tree.qza"
taxonomy_file <- "taxonomy.qza"
metadata_file <- "metadata.tsv"

#devtools::source_gist("8d0ca4206a66be7ff6d76fc4ab8e66c6")

cycle_1 <- qza_to_phyloseq(table_dada2,rooted_tree,taxonomy_file,metadata_file) 
cycle_1_amp <- phyloseq_to_ampvis2(cycle_1)

by_sample <- ampvis2::amp_rankabundance(cycle_1_amp, group_by = "SampleID") + theme(text=element_text(size = 20), legend.position="none")
by_sample 
ggsave("rank_abundance_curve_by_sample.png", plot=by_sample, path=outdir)

cycle_1_amp$metadata[[ioi]] <- factor(cycle_1_amp$metadata[[ioi]], levels=ioi_ord)

by_group <- ampvis2::amp_rankabundance(cycle_1_amp, group_by = ioi) + theme(text=element_text(size = 20))
by_group
ggsave("rank_abundance_curve_by_treatemnt.png", plot=by_group, path=outdir)
```